README.md
![Dutchie Logo](https://dutchie.com/favicons/default/apple-touch-icon.png)

# Test Rail Dart

This package is a thin wrapper around the Test Rail API that will allow for automated test reporting in Dart. It enables a user to start, report case pass/fail, and close your test runs from a Dart interface.

## Getting Started

Initialize the TestRail instance using the config method:

```dart
TestRail.config(
  username: 'USERNAME',
  password: 'PASSWORD',
  /// The url that points to the test rail server => https://example.testrail.com
  serverDomain: 'https://YOUR_SERVER.testrail.com'
)
```

## Usage

### Create or Update Runs

```dart
/// Start by creating a new run
final newRun = await TestRun.create(
  name: 'Test execution',
  projectId: 1
);

/// Add cases to the run
await newRun.updateRun(
  caseIds: [1, 2, 3, 5],
);
```

Once the run is created, results can be reported by case:

```dart
final result = await newRun.addResultForCase(
  caseId: 1,
  statusId: 1,
);

// Optionally add a screenshot or other image to the result
await result.addAttachmentToResult(
  '/workspace/attachments/failure.png',
);
```

### Get

Historical runs, cases, and sections can be retrieved:

```dart
final testCase = await TestCase.get(1);

final testRun = await TestRun.get(1);

final testSection = await TestSection.get(1);
```

Completed or ongoing test run results can be retrieved:

```dart
final caseResults = await TestResult.getCaseResults(
  runId: 1833,
  // Case ID is from TestCases, not TestRun
  caseId: 184234,
);

final runResults = await TestResult.getRunResults(
  runId: 1818,
  statusId: [5, 1],
);

final testResults = await TestResult.getTestResults(
  // Test ID from particular TestRun
  testId: 1868150,
);
```

## About Dutchie

We’re not just building the future of shopping for cannabis, we’re building a culture of innovation, customer care, and challenge to the status quo.

Inspired? Join our team of [Dart and Flutter developers](https://dutchie.com/careers) today

CHANGELOG.md
## 0.0.4
Update broken plan json data

## 0.0.3

### Fixed
*  Status id as correct parameter [#5](https://github.com/GetDutchie/test_rail_dart/pull/5)
## 0.0.2

* Update README
## 0.0.1

* Initial

lib/test_run.dart
import 'package:test_rail_dart/test_result.dart';
import 'package:test_rail_dart/src/test_rail_http_client.dart';
import 'package:test_rail_dart/test_rail.dart';
import 'package:test_rail_dart/test_runs.dart';

class TestRun {
  final int? assignedtoId;
  final int? blockedCount;
  final DateTime? completedOn;
  final String? config;
  final List<int>? configIds;
  final int? createdBy;
  final DateTime? createdOn;
  final int? customStatus1Count;
  final int? customStatus2Count;
  final int? customStatus3Count;
  final int? customStatus4Count;
  final int? customStatus5Count;
  final int? customStatus6Count;
  final int? customStatus7Count;
  final String? description;
  final int? failedCount;
  final int id;
  final bool? includeAll;
  final bool? isCompleted;
  final int? milestoneId;
  final String? name;
  final int? passedCount;
  final int? planId;
  final int? projectId;
  final String? refs;
  final int? retestCount;
  final int? suiteId;
  final int? untestedCount;
  final DateTime? updatedOn;
  final String? url;

  TestRun({
    this.assignedtoId,
    this.blockedCount,
    this.completedOn,
    this.config,
    this.configIds,
    this.createdBy,
    this.createdOn,
    this.customStatus1Count,
    this.customStatus2Count,
    this.customStatus3Count,
    this.customStatus4Count,
    this.customStatus5Count,
    this.customStatus6Count,
    this.customStatus7Count,
    this.description,
    this.failedCount,
    required this.id,
    this.includeAll,
    this.isCompleted,
    this.milestoneId,
    this.name,
    this.passedCount,
    this.planId,
    this.projectId,
    this.refs,
    this.retestCount,
    this.suiteId,
    this.untestedCount,
    this.updatedOn,
    this.url,
  });

  factory TestRun.fromJson(Map<String, dynamic> json) {
    DateTime? completedOn;
    DateTime? createdOn;
    DateTime? updatedOn;

    if (json['completed_on'] != null) {
      completedOn = DateTime.fromMillisecondsSinceEpoch(json['completed_on']);
    }

    if (json['created_on'] != null) {
      createdOn = DateTime.fromMillisecondsSinceEpoch(json['created_on']);
    }

    if (json['updated_on'] != null) {
      updatedOn = DateTime.fromMillisecondsSinceEpoch(json['updated_on']);
    }

    return TestRun(
      assignedtoId: json['assignedto_id'],
      blockedCount: json['blocked_count'],
      completedOn: completedOn,
      config: json['config'],
      configIds: json['config_ids']?.cast<int>(),
      createdBy: json['created_by'],
      createdOn: createdOn,
      customStatus1Count: json['custom_status1_count'],
      customStatus2Count: json['custom_status2_count'],
      customStatus3Count: json['custom_status3_count'],
      customStatus4Count: json['custom_status4_count'],
      customStatus5Count: json['custom_status5_count'],
      customStatus6Count: json['custom_status6_count'],
      customStatus7Count: json['custom_status7_count'],
      description: json['description'],
      failedCount: json['failed_count'],
      id: json['id'],
      includeAll: json['include_all'],
      isCompleted: json['is_completed'],
      milestoneId: json['milestone_id'],
      name: json['name'],
      passedCount: json['passed_count'],
      planId: json['plan_id'],
      projectId: json['project_id'],
      refs: json['refs'],
      retestCount: json['retest_count'],
      suiteId: json['suite_id'],
      untestedCount: json['untested_count'],
      updatedOn: updatedOn,
      url: json['url'],
    );
  }

  Future<TestResult> addResultForCase(
    int caseId, {
    String? comment,
    // in `1m 38s` format
    String? elapsed,
    required int statusId,
  }) async {
    final response = await TestRail.instance.client.request(
      '/add_result_for_case/$id/$caseId',
      RequestMethod.post,
      params: {
        'status_id': statusId,
        if (comment != null) 'comment': comment,
        if (elapsed != null) 'elapsed': elapsed,
      },
    );

    return TestResult.fromJson(response);
  }

  Future<TestRun> close() async {
    final response = await TestRail.instance.client
        .request('/close_run/$id', RequestMethod.post);
    return TestRun.fromJson(response);
  }

  static Future<TestRun> create({
    Iterable<int>? caseIds,
    String? description,
    bool includeAll = false,
    required String name,
    required int projectId,
  }) async {
    final response = await TestRail.instance.client.request(
      '/add_run/$projectId',
      RequestMethod.post,
      params: <String, dynamic>{
        if (caseIds != null) 'case_ids': caseIds,
        if (description != null) 'description': description,
        'include_all': includeAll,
        'name': name,
      },
    );
    return TestRun.fromJson(response);
  }

  static Future<TestRun> get(int runId) async {
    final response = await TestRail.instance.client
        .request('/get_run/$runId', RequestMethod.get);
    return TestRun.fromJson(response);
  }

  static Future<TestRuns> getAll({
    required int projectId,
    DateTime? createdAfter,
    DateTime? createdBefore,
    Iterable<int>? createdBy,
    bool? isCompleted,
    int? limit,
    int? offset,
    Iterable<int>? milestoneId,
    String? refsFilter,
    Iterable<int>? suiteId,
  }) async {
    final queryParameters = <String, dynamic>{
      'created_after': createdAfter?.millisecondsSinceEpoch,
      'created_before': createdBefore?.millisecondsSinceEpoch,
      'created_by': createdBy?.join(','),
      'is_completed': isCompleted,
      'limit': limit,
      'offset': offset,
      'milestone_id': milestoneId?.join(','),
      'refs_filter': refsFilter,
      'suite_id': suiteId?.join(','),
    };

    queryParameters.removeWhere((_, dynamic value) => value == null);

    final response = await TestRail.instance.client.request(
      '/get_runs/$projectId',
      RequestMethod.get,
      queryParameters: queryParameters,
    );

    return TestRuns.fromJson(response);
  }

  Future<TestRun> update({
    required Iterable<int> caseIds,
    bool includeAll = false,
  }) async {
    final response = await TestRail.instance.client.request(
      '/update_run/$id',
      RequestMethod.post,
      params: {
        'case_ids': caseIds,
        'include_all': includeAll,
      },
    );
    return TestRun.fromJson(response);
  }

  Map<String, dynamic> get asJson => {
        'assignedto_id': assignedtoId,
        'blocked_count': blockedCount,
        'completed_on': completedOn?.millisecondsSinceEpoch,
        'config_ids': configIds,
        'config': config,
        'created_by': createdBy,
        'created_on': createdOn?.millisecondsSinceEpoch,
        'custom_status1_count': customStatus1Count,
        'custom_status2_count': customStatus2Count,
        'custom_status3_count': customStatus3Count,
        'custom_status4_count': customStatus4Count,
        'custom_status5_count': customStatus5Count,
        'custom_status6_count': customStatus6Count,
        'custom_status7_count': customStatus7Count,
        'description': description,
        'failed_count': failedCount,
        'id': id,
        'include_all': includeAll,
        'is_completed': isCompleted,
        'milestone_id': milestoneId,
        'name': name,
        'passed_count': passedCount,
        'plan_id': planId,
        'project_id': projectId,
        'refs': refs,
        'retest_count': retestCount,
        'suite_id': suiteId,
        'untested_count': untestedCount,
        'updated_on': updatedOn?.millisecondsSinceEpoch,
        'url': url,
      };
}


lib/test_case.dart
import 'package:test_rail_dart/src/test_rail_http_client.dart';
import 'package:test_rail_dart/test_rail.dart';

class TestCase {
  final int? createdBy;
  final DateTime? createdOn;
  final String? customExpected;
  final String? customPreconds;
  final String? customSteps;
  final List<CustomStep>? customStepsSeparated;
  final String? estimate;
  final String? estimateForecast;
  final int id;
  final int? milestoneId;
  final int? priorityId;
  final String? refs;
  final int? sectionId;
  final int? suiteId;
  final String? title;
  final int? typeId;
  final int? updatedBy;
  final DateTime? updatedOn;

  TestCase({
    this.createdBy,
    this.createdOn,
    this.customExpected,
    this.customPreconds,
    this.customSteps,
    this.customStepsSeparated,
    this.estimate,
    this.estimateForecast,
    required this.id,
    this.milestoneId,
    this.priorityId,
    this.refs,
    this.sectionId,
    this.suiteId,
    this.title,
    this.typeId,
    this.updatedBy,
    this.updatedOn,
  });

  factory TestCase.fromJson(Map<String, dynamic> json) {
    DateTime? createdOn;
    DateTime? updatedOn;

    if (json['created_on'] != null) {
      createdOn = DateTime.fromMillisecondsSinceEpoch(json['created_on']);
    }

    if (json['updated_on'] != null) {
      updatedOn = DateTime.fromMillisecondsSinceEpoch(json['updated_on']);
    }
    var customStepsSeparated = <CustomStep>[];

    if (json['custom_steps_separated'] != null) {
      customStepsSeparated = List.from(
        json['custom_steps_separated'].map(
          (stepJson) => CustomStep.fromJson(stepJson),
        ),
      );
    }

    return TestCase(
      createdBy: json['created_by'],
      createdOn: createdOn,
      customExpected: json['custom_expected'],
      customPreconds: json['custom_preconds'],
      customSteps: json['custom_steps'],
      customStepsSeparated: customStepsSeparated,
      estimate: json['estimate'],
      estimateForecast: json['estimate_forecast'],
      id: json['id'],
      milestoneId: json['milestone_id'],
      priorityId: json['priority_id'],
      refs: json['refs'],
      sectionId: json['section_id'],
      suiteId: json['suite_id'],
      title: json['title'],
      typeId: json['type_id'],
      updatedBy: json['updated_by'],
      updatedOn: updatedOn,
    );
  }

  static Future<TestCase> get(int caseId) async {
    final response = await TestRail.instance.client
        .request('/get_case/$caseId', RequestMethod.get);
    return TestCase.fromJson(response);
  }

  Map<String, dynamic> get asJson => {
        'created_by': createdBy,
        'created_on': createdOn?.millisecondsSinceEpoch,
        'custom_steps_separated':
            customStepsSeparated?.map((v) => v.asJson).toList(),
        'custom_expected': customExpected,
        'custom_preconds': customPreconds,
        'custom_steps': customSteps,
        'estimate': estimate,
        'estimate_forecast': estimateForecast,
        'id': id,
        'milestone_id': milestoneId,
        'priority_id': priorityId,
        'refs': refs,
        'section_id': sectionId,
        'suite_id': suiteId,
        'title': title,
        'type_id': typeId,
        'updated_by': updatedBy,
        'updated_on': updatedOn?.millisecondsSinceEpoch,
      };
}

class CustomStep {
  final String? content;
  final String? additionalInfo;
  final String? expected;
  final String? refs;
  final int? sharedStepId;

  CustomStep({
    this.content,
    this.additionalInfo,
    this.expected,
    this.refs,
    this.sharedStepId,
  });

  factory CustomStep.fromJson(Map<String, dynamic> json) {
    return CustomStep(
      content: json['content'],
      additionalInfo: json['additional_info'],
      expected: json['expected'],
      refs: json['refs'],
      sharedStepId: json['shared_step_id'],
    );
  }

  Map<String, dynamic> get asJson => {
        'content': content,
        'additional_info': additionalInfo,
        'expected': expected,
        'refs': refs,
        if (sharedStepId != null) 'shared_step_id': sharedStepId,
      };
}


lib/test_rail.dart
import 'package:http/http.dart' as http;
import 'package:test_rail_dart/src/test_rail_http_client.dart';

class TestRail {
  final TestRailHttpClient client;

  TestRail._({
    required this.client,
  });

  static late TestRail instance;

  static TestRail configure({
    required String password,

    /// The url that points to the TestRail server => https://example.testrail.com
    required String serverDomain,
    required String username,
    http.Client? client,
  }) {
    return instance = TestRail._(
      client: TestRailHttpClient(
        client: client,
        password: password,
        serverDomain: serverDomain,
        username: username,
      ),
    );
  }
}


example/main.dart
// ignore_for_file: unused_local_variable
import 'package:test_rail_dart/test_rail.dart';
import 'package:test_rail_dart/test_case.dart';
import 'package:test_rail_dart/test_result.dart';
import 'package:test_rail_dart/test_run.dart';
import 'package:test_rail_dart/test_section.dart';

void main(List<String> args) async {
  TestRail.configure(
    serverDomain: args[0],
    username: args[1],
    password: args[2],
  );

  // replace with your own caseId
  final testCase = await TestCase.get(142864);

  // replace with your own projectId, extra parameters available in method
  final testRuns = await TestRun.getAll(projectId: 1125, limit: 10);

  // replace with your own projectId
  final testRun = await TestRun.get(1125);

  // replace with your own sectionId
  final section = await TestSection.get(13);
  final newRun = await TestRun.create(
    caseIds: [testCase.id],
    description: '121',
    name: 'newRun',
    // replace with your own projectId
    projectId: 134,
  );

  final testCaseResult = await newRun.addResultForCase(
    testCase.id,
    // Status - 1: PASSED
    statusId: 1,
  );

  await newRun.update(
    caseIds: [1, 6, 7],
    includeAll: false,
  );

  // Replace with attachment path
  await testCaseResult.addAttachment('<attachment_path>');

  final caseResults = await TestResult.getCaseResults(
    // Case ID is from TestCases, not TestRun
    184234,
    runId: 1833,
  );

  final runResults = await TestResult.getRunResults(
    1818,
    statusId: [5, 1],
  );

  final testResults = await TestResult.getTestRunSingleTestResults(
    // Test ID from particular TestRun
    testId: 1868150,
  );
}
